demo.py

```diff
@@ -0,0 +1,146 @@
+class Resource(models.Model):
+    client = ForeignKey(Client)
+    owned_by = ForgeignKey(Resource, null=True)
+    name = CharField(blank=False, max_length=128)
+    configuration = JSONField(null=False)
+    types = ResourceTypeRegistry()
+    owned_resources = ResourceManger()
+
+    @property
+    def configuration_schema(self):
+        """ Define the options used to configure this resource. """
+        return {}  # No configurable options by default
+
+    @property
+    def state(self):
+        """
+        The current state of this resource.
+        The return value is always a subclass of ResourceState, and typically will be a subclass of:
+          NeedsRefreshState, ApplyingConfigurationState, NeedsInputState, ReadyState, ErrorState, or DestroyingState
```

Comment:
I see, interesting - I've never used state machine before, so sorry for the newbie comments : )

Here, does that mean that, for example, NeedRefresh, ReadyState & ErrorState would actually need to be turned into methods on other states? I would be curious to see how the current states in the app match against a state machine.